src/software/ai/hl/stp/action/intercept_ball_action.cpp

```diff
@@ -0,0 +1,123 @@
+#include "software/ai/hl/stp/action/intercept_ball_action.h"
+
+#include "shared/constants.h"
+#include "software/ai/evaluation/pass.h"
+#include "software/ai/hl/stp/evaluation/intercept.h"
+#include "software/ai/hl/stp/evaluation/robot.h"
+#include "software/ai/intent/move_intent.h"
+#include "software/geom/ray.h"
+#include "software/geom/util.h"
+#include "software/util/logger/init.h"
+
+InterceptBallAction::InterceptBallAction(const Field& field, const Ball& ball,
+                                         bool loop_forever)
+    : Action(), field(field), ball(ball), loop_forever(loop_forever)
+{
+}
+
+std::unique_ptr<Intent> InterceptBallAction::updateStateAndGetNextIntent(
+    const Robot& robot, const Field& field, const Ball& ball)
+{
+    // Update the parameters stored by this Action
+    this->field = field;
+    this->ball  = ball;
+    this->robot = robot;
+
+    return getNextIntent();
+}
+
+std::optional<Point> InterceptBallAction::getPointBallLeavesField(const Field& field,
+                                                                  const Ball& ball)
+{
+    Ray ball_ray(ball.position(), ball.velocity());
+    if (field.pointInFieldLines(ball.position()))
+    {
+        return rayRectangleIntersection(ball_ray, field.fieldLines()).first;
+    }
+    else
+    {
+        return std::nullopt;
+    }
+}
+
+void InterceptBallAction::calculateNextIntent(IntentCoroutine::push_type& yield)
+{
+    do
+    {
+        Point closest_point = closestPointOnLine(robot->position(), ball.position(),
+                                                 ball.position() + ball.velocity());
+        bool point_in_front_of_ball =
+            pointInFrontVector(ball.position(), ball.velocity(), closest_point);
+
+        Duration ball_time_to_position = Duration::fromSeconds(
+            dist(closest_point, ball.position()) / (ball.velocity().len() + 1e-6));
```

Comment:
The case where the ball is not moving should be handled explicitly

Reply:
I don't entirely agree. Would you prefer the case to be handled explicitly because it's more clear?

It results in the same behavior but with a guarantee of avoiding division by 0, but a ball takes an 'infinitely long time' to go somewhere if ti's basically stationary. IMO another `if` block with another case for the ball being stationary doesn't necessarily make the logic easier to follow (i'd like to minimize conditional blocks if possible). I can at least comment to indicate we are doing the addition to avoid division by 0.

Thoughts?

Reply:
Using a constant with an explicit comment works for me. 

It's just not clear (to me at least when I saw it) that you are avoiding /0 instead of some sort of offset.

I'm also pretty sure if ball velocity is 0 'pointInFrontVector()' will not return true for 50% of positions

Reply:
I'd prefer explicit handling

Reply:
I'd rather have explicit handling here, not a blocker

Reply:
This action will be re-written significantly later (since ball interception & collection is one of our high-level goals), so lets just deal with this then.